core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -489,14 +479,11 @@ public synchronized ListenableFuture<List<TaskDescriptor>> getMoreTasks()
                             Map<Integer, ListMultimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
                             SetMultimap<Integer, HostAddress> partitionToNodeMap = HashMultimap.create();
                             for (LoadedSplits loadedSplits : loadedSplitsList) {
-                                BucketNodeMap bucketNodeMap = this.bucketNodeMap
-                                        .orElseThrow(() -> new VerifyException("bucket to node map is expected to be present"));
                                 for (Split split : loadedSplits.getSplits()) {
-                                    int bucket = bucketNodeMap.getBucket(split);
-                                    int partition = getPartitionForBucket(bucket);
-
-                                    {
-                                        HostAddress requiredAddress = bucketNodeMap.getAssignedNode(split).getHostAndPort();
+                                    int partition = sourcePartitioningScheme.getPartition(split);
+                                    Optional<InternalNode> assignedNode = sourcePartitioningScheme.getNodeRequirement(partition);
+                                    if (assignedNode.isPresent()) {
```

Comment:
why do we need this check now?

Reply:
Currently the buckets are always pre-assigned to nodes (this is how `NodePartitioningManager` works). This is a problem for fault tolerant execution as it limits the number of partitions (sets it equal to the number of nodes). In the future we need to resolve this issue. Once the `NodePartitioningManager` no longer pre-assignes the nodes to buckets there will be no more partition-to-node mapping available. I wanted to make it future proof to avoid another refactoring.